File: solutions.digamma.damas.standalone/src/test/java/solutions/digamma/damas/jcr/providers/TestRepositoryFactory.java
package solutions.digamma.damas.jcr.providers;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import javax.annotation.PreDestroy;
import javax.annotation.Priority;
import javax.decorator.Decorator;
import javax.decorator.Delegate;
import javax.inject.Inject;
import javax.interceptor.Interceptor;
import javax.jcr.RepositoryFactory;
import solutions.digamma.damas.config.Configuration;
import solutions.digamma.damas.config.Fallback;

/**
 * Repository factory decorator that drops repository upon shutdown.
 */
@Decorator
@Priority(Interceptor.Priority.APPLICATION)
abstract class TestRepositoryFactory implements RepositoryFactory {

    @Delegate
    @Inject
    private RepositoryFactory ignore;

    @Inject
    @Configuration("repository.home")
    @Fallback("storage")
    private String repositoryHome;

    @PreDestroy
    public void cleanUp() throws IOException {
        Files.walk(Paths.get(this.repositoryHome))
                .map(Path::toFile)
                .forEach(File::deleteOnExit);
    }
}

File: solutions.digamma.damas.jcr/src/test/kotlin/solutions/digamma/damas/jcr/TestRepositoryFactory.kt
package solutions.digamma.damas.jcr

import solutions.digamma.damas.config.Configuration
import solutions.digamma.damas.config.Fallback
import solutions.digamma.damas.logging.Logbook
import java.io.File
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import javax.annotation.PreDestroy
import javax.annotation.Priority
import javax.decorator.Decorator
import javax.decorator.Delegate
import javax.inject.Inject
import javax.interceptor.Interceptor
import javax.jcr.RepositoryFactory

@Decorator
@Priority(Interceptor.Priority.APPLICATION)
abstract class TestRepositoryFactory: RepositoryFactory {

    @Delegate
    @Inject
    private lateinit var ignore: RepositoryFactory

    @Inject
    private lateinit var log: Logbook

    @Inject
    @Configuration("repository.home")
    @Fallback("storage")
    private lateinit var repositoryHome: String

    @PreDestroy
    fun cleanUp() {
        this.log.info("Deleting files at ${this.repositoryHome}.")
        Files.walk(Paths.get(this.repositoryHome))
            .sorted(Comparator.reverseOrder())
            .map { it.toFile() }
            .forEach { it.delete() }
            this.log.info("Repository home purged.")
    }
}
